### ISSUE TITLE
Translating Atomic stdlib functions correctly

### ISSUE THREAD

**Issue by anmolsahoo25** [ocaml-multicore/ocaml-multicore#254]
Currently, the functions from the `Atomic` library are translated as field accesses into module blocks by the compiler. This way, the `Atomic.get` function is always translated as a conservative load to a pointer. 

We want to specialize the `Atomic.get` depending on the value being loaded, where loads to immediate and pointer values are generated correctly.

**Comment by kayceesrk** [comment]
Related to PR #251 and issue #237.

**Comment by kayceesrk** [comment]
@stedolan Any thoughts on how to do this right?

**Comment by stedolan** [comment]
To optimise this, you need to add a case to `Translcore.specialize_primitive`, following the same pattern as `Psetfield`

**Comment by anmolsahoo25** [comment]
The issue with this is that `Atomic.make` gets translated as a `Val_reg` from the syntax tree. Thus, the `specialize_primitive` function is not called with `Atomic.make` as an input.

I can pull up the function definitions and execution steps, if required. I inspected this behavior using gdb. I could be wrong here, if needed, I can revisit this process.

**Comment by anmolsahoo25** [comment]
Just adding this here for future reference and my own understanding since the other participants would already be aware of these details - 

These are the two snippets of code, I am trying to compile -

```ocaml
let x = Atomic.make 0
```

```ocaml
let x = Pervasives.ref 0
```

Now dumping the TypedTree for both of them (omitted for clarity) -

```
...
            Texp_ident "Pervasives!.ref"
...
```
```
...
            Texp_ident "Atomic!.make"
...
```

Now dumping the raw lambda - 

```
(seq
  (let (x/1002 = (makemutable 0 (int) 0))
    (setfield_ptr(root-init) 0 (global Test_make!) x/1002))
  0a)
```

```
(seq
  (let (x/1002 = (apply (field_imm 0 (global Atomic!)) 0))
    (setfield_ptr(root-init) 0 (global Atomic_make!) x/1002))
  0a)
```

Now clearly, the idents got translated differently. We can see this here, thus `specialize_primitive` is not called for `Val_reg` - 
https://github.com/ocaml-multicore/ocaml-multicore/blob/8fbacb11cc91d6f8d5a64a7b335a8c2d97480e52/bytecomp/translcore.ml#L685

https://github.com/ocaml-multicore/ocaml-multicore/blob/8fbacb11cc91d6f8d5a64a7b335a8c2d97480e52/bytecomp/translcore.ml#L706

The translation from `Pexp_ident` to `Texp_ident` happens here - 
https://github.com/ocaml-multicore/ocaml-multicore/blob/8fbacb11cc91d6f8d5a64a7b335a8c2d97480e52/typing/typecore.ml#L2706

This implies that `Pexp_ident` has been tagged as `Val_reg` or `Val_prim` before typing. This description comes from this function in `env.ml` - 
https://github.com/ocaml-multicore/ocaml-multicore/blob/8fbacb11cc91d6f8d5a64a7b335a8c2d97480e52/typing/env.ml#L1212

These components are added here - 
https://github.com/ocaml-multicore/ocaml-multicore/blob/8fbacb11cc91d6f8d5a64a7b335a8c2d97480e52/typing/env.ml#L1671

**Comment by anmolsahoo25** [comment]
Thus, it has to be figured as to how to add the elements of the module `Atomic` as primitives. This would then correctly translate the elements as `Val_prim` and then entries can be made in `specialize_primitive`.

I could be completely wrong here! So, I am still investigating.

**Comment by anmolsahoo25** [comment]
I figured out the issue. Unfortunately, the ```atomic.mli``` file had the declarations for `make` as - 

```
val make : 'a -> 'a t = "%makemutable"
```

which should have been
```
external make : 'a -> 'a t = "%makemutable"
```

Not declaring the value as external in the `mli` file, causes it be to be generated as a field access.

**Comment by anmolsahoo25** [comment]
Closed by #262
